elements/pl-dataframe/pl-dataframe.py

```diff
@@ -0,0 +1,164 @@
+from enum import Enum
+from typing import cast
+
+import chevron
+import lxml.html
+import pandas as pd
+import prairielearn as pl
+from typing_extensions import assert_never
+
+
+class DisplayLanguage(Enum):
+    PYTHON = 1
+    R = 2
+
+
+SHOW_HEADER_DEFAULT = True
+SHOW_INDEX_DEFAULT = True
+SHOW_DIMENSIONS_DEFAULT = True
+DISPLAY_LANGUAGE_DEFAULT = DisplayLanguage.PYTHON
+SHOW_DTYPE_DEFAULT = False
+NUM_DIGITS_DEFAULT = None
+SHOW_PYTHON_DEFAULT = True
+
+
+def convert_pandas_dtype_to_r(s: pd.Series) -> str:
+    # Force series to avoid odd element-wise output
+    s.dtype
+
+    if pd.api.types.is_float_dtype(s):
+        return "numeric"
+    elif pd.api.types.is_integer_dtype(s):
+        return "integer"
+    elif pd.api.types.is_object_dtype(s) or pd.api.types.is_string_dtype(s):
+        return "character"
+    elif pd.api.types.is_categorical_dtype(s):
+        # Check if ordered
+        if s.cat.ordered:
+            return "ordered factor"
+        return "factor"
+    elif pd.api.types.is_bool_dtype(s):
+        return "logical"
+    elif pd.api.types.is_complex_dtype(s):
+        return "complex"
+    elif pd.api.types.is_datetime64_any_dtype(s):
+        return "POSIXct"
+    elif pd.api.types.is_timedelta64_dtype(s) or pd.api.types.is_period_dtype(s):
+        return "Not supported"
+
+    return "Unknown"
+
+
+def get_pandas_dtype(s: pd.Series) -> str:
+    return s.dtype
+
+
+def prepare(element_html: str, data: pl.QuestionData) -> None:
+    element = lxml.html.fragment_fromstring(element_html)
+    pl.check_attribs(
+        element,
+        required_attribs=["params-name"],
+        optional_attribs=[
+            "show-index",
+            "show-header",
+            "show-dimensions",
+            "show-dtype",
+            "show-python",
+            "display-language",
+            "digits",
+        ],
+    )
+
+
+def render(element_html: str, data: pl.QuestionData) -> str:
+    element = lxml.html.fragment_fromstring(element_html)
+    # TODO should there be a different display varname?
+    varname = pl.get_string_attrib(element, "params-name")
```

Comment:
Any reason not to just render this as `df = pd.DataFrame(...)`? I recognize there's precedence for reusing the parmas name for a user-facing variable name (e.g. in `pl-matrix-output`), but I don't necessarily agree that that's a good idea. A key in `params` may not necessarily be a valid Python identifier and it also may have no relation to what students should see.

Reply:
Will defer to @coatless on what they think, but always using `df` seems reasonable to me.

Reply:
I think the `varname` component came into play to mimic being able to set a variable name in `pl-variable-output`, c.f. `<variable params-name=""></variable>` tag:

https://prairielearn.readthedocs.io/en/latest/elements/#pl-variable-output-element

I'm not opposed to defaulting to `df` as the internal variable. 

Though, it would be nice to provide an option to specify a data frame output name. This would help with providing flavor text to the problem. For example:

```
"Please use the `employees` data frame to find the employee with the highest salary" 
```

vs.

```
"Please use the `df` data frame
```

Reply:
I'll go ahead and add an option that defaults to `df` then, since it's pretty straightforward.

Reply:
Added the ability to change this with the `display-variable-name` attribute.